hivemind/optim/averaged.py

```diff
@@ -105,6 +172,54 @@ def _average_parameters_in_background(
             except Exception as e:
                 logger.error(f"Averaging round failed: caught {e}.")
 
+    def is_synchronized(self) -> bool:
+        return self.local_epoch + self.max_allowed_epoch_difference >= self.training_state.max_epoch
+
+    @torch.no_grad()
+    def _report_progress(self):
+        while not self.stop_event.is_set(): 
+            self.report_progress_event.wait()
+            self.report_progress_event.clear()
+            if self.stop_event.is_set():
+                break
+            current_time = get_dht_time()
+            with self.lock_scheduler_params:
+                local_state_info = [self.local_step, current_time, self.local_epoch]
```

Comment:
Convert this to a dataclass with validation, same as hivemind.optim.collaborative.TrainingState